src/wikirec/model.py

```diff
@@ -327,3 +351,134 @@
+            # Add index of book and index of link to pairs
+            if link.lower() in links:
+                pairs.append((book_index[title], link_index[link.lower()]))
+    pairs_set = set(pairs)
+    
+    # Neural network architecture
+    # Both inputs are 1-dimensional
+    book_input = Input(name = 'book', shape = [1])
+    link_input = Input(name = 'link', shape = [1])
+    
+    # Embedding the book (shape will be (None, 1, 50))
+    book_embedding = Embedding(name = 'book_embedding',
+                               input_dim = len(book_index),
+                               output_dim = embedding_size)(book_input)
+    
+    # Embedding the link (shape will be (None, 1, 50))
+    link_embedding = Embedding(name = 'link_embedding',
+                               input_dim = len(link_index),
+                               output_dim = embedding_size)(link_input)
+    
+    # Merge the layers with a dot product along the second axis 
+    # (shape will be (None, 1, 1))
+    merged = Dot(name = 'dot_product', normalize = True, 
+                 axes = 2)([book_embedding, link_embedding])
+    
+    # Reshape to be a single number (shape will be (None, 1))
+    merged = Reshape(target_shape = [1])(merged)
+    
+    model = Model(inputs = [book, link], outputs = merged)
+    model.compile(optimizer = 'Adam', loss = 'mse')
+    
+    # Function that creates a generator for training data 
+    def _generate_batch(pairs, n_positive = 50, negative_ratio = 1.0):
+        """Generate batches of samples for training. 
+           Random select positive samples
+           from pairs and randomly select negatives."""
+
+        # Create empty array to hold batch
+        batch_size = n_positive * (1 + negative_ratio)
+        batch = np.zeros((batch_size, 3))
+
+        # Continue to yield samples
+        while True:
+            # Randomly choose positive examples
+            for idx, (book_id, link_id) in enumerate(random.sample(pairs, n_positive)):
+                batch[idx, :] = (book_id, link_id, 1)
+            idx += 1
+
+            # Add negative examples until reach batch size
+            while idx < batch_size:
+
+                # Random selection
+                random_book = random.randrange(len(book_index))
+                random_link = random.randrange(len(link_index))
+
+                # Check to make sure this is not a positive example
+                if (random_book, random_link) not in pairs_set:
+
+                    # Add to batch and increment index
+                    batch[idx, :] = (random_book, random_link, 0)
+                    idx += 1
+
+            # Make sure to shuffle order
+            np.random.shuffle(batch)
+            yield {'book': batch[:, 0], 'link': batch[:, 1]}, batch[:, 2]
+
+    n_positive = 1024
+    gen = _generate_batch(pairs, n_positive, negative_ratio = 2)
+    h = model.fit_generator(gen, epochs = 15, steps_per_epoch = len(pairs) // n_positive)
```

Comment:
Just checking here, h is not used as it's just a placeholder for the model being fit. I'd add a pylint: disable=unused-variable if that's the case :)

Reply:
@andrewtavis Yep! h is just a placeholder.